math/sieve_of_eratosthenes.cpp

```diff
@@ -10,24 +10,18 @@
  * @see primes_up_to_billion.cpp prime_numbers.cpp
  */
 
-#include <iostream>
-
-/** Maximum number of primes */
-#define MAX 10000000
-
-/** array to store the primes */
-bool isprime[MAX];
+#include <iostream> // for io operations
 
 /**
  * This is the function that finds the primes and eliminates
  * the multiples.
  */
-void sieve(uint32_t N) {
+void sieve(uint32_t N, bool *isprime) {
     isprime[0] = false;
     isprime[1] = false;
-    for (uint32_t i = 2; i <= N; i++) {
+    for (uint32_t i = 2; i * i <= N; i++) {
         if (isprime[i]) {
-            for (uint32_t j = (i << 1); j <= N; j += i) {
+            for (uint32_t j = i * i; j <= N; j = j + i) {
```

Comment:
The original implementation was correct. 
* `i<<1` -> this means, start from `i*2` and mark all multiples of i as not-prime
* What you have done: `i*i` -> this means, start from `i*i` and mark all multiples of i as not-prime. Thus, you have lost all numbers from `i*2` to `i*i`!

Reply:
@kvedala I am not getting the difference between using `i << 1` and `i * i` since I am getting same output for both the cases when I executed the program locally.

Reply:
@kvedala I am not getting the difference between using `i << 1` and `i * i` since I am getting the same output for both the cases when I executed the program locally.

Reply:
the difference is as mentioned above:
1. the inner loop start is `i^2` (i squared) in your case and `i*2` (i times 2) in the original implementation
2. In the original implementation, the outer loop will skip for many numbers less than `i*i` because the inner loop will mark the non-primes. Hence, the inner loop will not execute for all numbers
3. In your case, he outer loop will check every number less then i squared and hence, the inner loop will execute for every number!

If it is still not clear, try to work the loop for i 23 on a piece of paper using the above algorithm step by step or use cout statements and check every possibility.

Reply:
Oh I got it you mean for example I = 2 it will skip 2 and start from 4 according to my code is it right @kvedala

Reply:
Oh, I got it you mean for example i = 2 it will skip 2 and start from 4 according to my code is it right @kvedala

Reply:
yes, @iamrajiv

Reply:
make this fix as well.

Reply:
@kvedala I am not getting it should I revert the code to `for (uint32_t j = (i << 1); j <= N; j += i)`

Reply:
yes